### PR TITLE
[MRG] New API should allow prediction functions and scoring

### PR DESCRIPTION
**Pull request by wdevazelhes**
The new API should allow metric learning algorithms that fit on tuples of points to also predict, score, etc on tuples, Therefore being usable in scikit-learn's cross-validation routines. This is part of the PRs need for issue #91.

- [x] Take previous tests from https://github.com/metric-learn/metric-learn/pull/85, and refactor them using pytest, and to allow already formed tuples (3D arrays) instead of ``ConstrainedDatasets``.
- [x] Take previous code from https://github.com/metric-learn/metric-learn/pull/85 and adapt it to formed tuples
- [x] Make tests work
- [x] Make some modifications if needed
    - <strike>Move docstrings from ``_fit`` to ``fit``</strike>
    - <strike>Remove unused imports</strike>

### CODE REVIEW

**Comment by wdevazelhes** [comment]
I just merged with the recently merged PR, so changes introduced by this PR are now clearer.

File: metric_learn/base_metric.py [hunk 1]
```diff
@@ -49,3 +54,105 @@ def transform(self, X=None):
       X = check_array(X, accept_sparse=True)
     L = self.transformer()
     return X.dot(L.T)
+
+
+class _PairsClassifierMixin:
+
+  def predict(self, pairs):
+    """Predicts the learned similarity between input pairs.
```
Comment (bellet) [inline metric_learn/base_metric.py]: should be metric instead of similarity here

File: metric_learn/base_metric.py [hunk 2]
```diff
@@ -49,3 +54,105 @@ def transform(self, X=None):
...
+    score : float
+      The ``roc_auc`` score.
+    """
+    return roc_auc_score(y, self.decision_function(pairs))
+
+
+class _QuadrupletsClassifierMixin:
+
+  def predict(self, quadruplets):
+    """Predicts differences between sample similarities in input quadruplets.
```
Comment (bellet) [inline metric_learn/base_metric.py]: distances?

File: test/test_weakly_supervised.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  num_constraints = 20
+  constraints = Constraints.random_subset(y)
+  pairs = constraints.positive_negative_pairs(num_constraints,
+                                              same_length=True,
+                                              random_state=RNG)
+  return X, pairs
+
+
+def build_pairs():
+  # test that you can do cross validation on a ConstrainedDataset with
```
Comment (bellet) [inline test/test_weakly_supervised.py]: no ConstrainedDataset anymore. also `X_constrained` should be renamed (this is a set of pairs)

File: test/test_weakly_supervised.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  X_constrained, y = wrap_pairs(X, pairs)
+  X_constrained, y = shuffle(X_constrained, y)
+  (X_constrained_train, X_constrained_test, y_train,
+   y_test) = train_test_split(X_constrained, y)
+  return (X_constrained, y, X_constrained_train, X_constrained_test,
+          y_train, y_test)
+
+
+def build_quadruplets():
+  # test that you can do cross validation on a ConstrainedDataset with
```
Comment (bellet) [inline test/test_weakly_supervised.py]: same here

File: metric_learn/base_metric.py [hunk 5]
```diff
@@ -49,3 +54,105 @@ def transform(self, X=None):
...
+    -------
+    prediction : np.ndarray of floats, shape=(n_constraints,)
+      Metric differences.
+    """
+    similar_diffs = quadruplets[:, 0, :] - quadruplets[:, 1, :]
+    dissimilar_diffs = quadruplets[:, 2, :] - quadruplets[:, 3, :]
+    return (np.sqrt(np.sum(similar_diffs.dot(self.metric()) *
+                           similar_diffs, axis=1)) -
+            np.sqrt(np.sum(dissimilar_diffs.dot(self.metric()) *
+                           dissimilar_diffs, axis=1)))
```
Comment (perimosocordiae) [inline metric_learn/base_metric.py]: This pattern, distance under some metric, seems like it should be factored out.

File: metric_learn/base_metric.py [hunk 6]
```diff
@@ -49,3 +54,105 @@ def transform(self, X=None):
...
+
+    y : Ignored, for scikit-learn compatibility.
+
+    Returns
+    -------
+    score : float
+      The quadruplets score.
+    """
+    predicted_sign = self.decision_function(quadruplets) < 0
+    return np.sum(predicted_sign) / predicted_sign.shape[0]
```
Comment (perimosocordiae) [inline metric_learn/base_metric.py]: Why not `np.mean(np.sign(...))` here?

File: metric_learn/base_metric.py (hunk 2 above)
Comment (wdevazelhes) [reply metric_learn/base_metric.py]: Yes, thanks

File: test/test_weakly_supervised.py (hunk 3 above)
Comment (wdevazelhes) [reply test/test_weakly_supervised.py]: Yes, thanks

File: metric_learn/base_metric.py (hunk 5 above)
Comment (wdevazelhes) [reply metric_learn/base_metric.py]: Yes indeed, the function will call function ``score_pairs`` (that returns the new metric between points) that will be inherited from the BaseMetricLearner, and implemented through ``ExplicitMixin`` (a Mixin for all learners that can embed data) (so ``score_pairs`` will be implemented as the euclidean distance between embeddings)

File: metric_learn/base_metric.py (hunk 6 above)
Comment (wdevazelhes) [reply metric_learn/base_metric.py]: Much cleaner indeed, thanks !

File: test/test_weakly_supervised.py (hunk 3 above)
Comment (wdevazelhes) [reply test/test_weakly_supervised.py]: as some tests are parameterized to work for pairs and quadruplets, I will rename them tuples in the tests, but pairs and quadruplets in ``build_pairs`` and ``build_quadruplets`` functions that initialize data

File: metric_learn/base_metric.py (hunk 5 above)
Comment (bellet) [reply metric_learn/base_metric.py]: (this should ultimately be in the Mahalanobis Mixin)
